controllers/cluster_controller.go

```diff
@@ -246,8 +252,26 @@ func (r *ClusterReconciler) upgradeControlPlane(ctx context.Context, cluster *ca
 			return false, errors.Wrapf(err, "failed to build infrastructure template from %q", kcp.Spec.InfrastructureTemplate.Name)
 		}
 
+		imagePath := strings.Split(KindToMachineImagePath[kcp.Spec.InfrastructureTemplate.Kind], ".")
+		for i := 1; i < len(imagePath); i++ {
+			parent := imagePath[0:i]
+			parentStr := strings.Join(parent, ".")
+			if _, found, err := unstructured.NestedFieldNoCopy(newInfrastructureMachineTemplate.Object, parent...); err != nil {
+				return false, errors.Wrapf(err, "failed to access `%s` in infrastructure template object %q", parentStr, newInfrastructureMachineTemplate.GetName())
+			} else if !found {
+				logger.Info("key not found in infrastructure machine template, defaulting", "infrastructureMachineTemplate", newInfrastructureMachineTemplate.GetName(), "key", parentStr)
+				if err = unstructured.SetNestedMap(newInfrastructureMachineTemplate.Object, map[string]interface{}{}, parent...); err != nil {
```

Comment:
can we add here the other fields like `publisher` or `offer?

Reply:
Sure, we can, the only question is where do we get the data from? Hardcoding it here would work for now, but it seems we need a more generic idea going forward, like having the entire image spec in configuration or in some resource or in the `Release`? Don't know, WDYT?